src/wayland/compositor/transaction.rs

```diff
@@ -0,0 +1,248 @@
+            inner: Arc::new(Mutex::new(TransactionInner::Data(Default::default()))),
+        }
+    }
+}
+
+impl PendingTransaction {
+    fn with_inner_state<T, F: FnOnce(&mut TransactionState) -> T>(&self, f: F) -> T {
+        let mut next = self.inner.clone();
+        loop {
+            let tmp = match *next.lock().unwrap() {
+                TransactionInner::Data(ref mut state) => return f(state),
+                TransactionInner::Fused(ref into) => into.clone(),
+            };
+            next = tmp;
+        }
+    }
+
+    pub(crate) fn insert_state(&self, surface: WlSurface, id: Serial) {
+        self.with_inner_state(|state| {
+            if let Some(place) = state.surfaces.iter_mut().find(|place| place.0 == surface) {
+                // the surface is already in the list, update the serial
+                if place.1 < id {
+                    place.1 = id;
+                }
+            } else {
+                // the surface is not in the list, insert it
+                state.surfaces.push((surface, id));
+            }
+        })
+    }
+
+    pub(crate) fn add_blocker<B: Blocker + Send + 'static>(&self, blocker: B) {
+        self.with_inner_state(|state| state.blockers.push(Box::new(blocker) as Box<_>))
+    }
+
+    pub(crate) fn is_same_as(&self, other: &PendingTransaction) -> bool {
+        let ptr1 = self.with_inner_state(|state| state as *const _);
+        let ptr2 = other.with_inner_state(|state| state as *const _);
+        ptr1 == ptr2
+    }
+
+    pub(crate) fn merge_into(&self, into: &PendingTransaction) {
+        if self.is_same_as(into) {
+            // nothing to do
+            return;
+        }
+        // extract our pending surfaces and change our link
+        let mut next = self.inner.clone();
+        let my_state;
+        loop {
+            let tmp = {
+                let mut guard = next.lock().unwrap();
+                match *guard {
+                    TransactionInner::Data(ref mut state) => {
+                        my_state = std::mem::take(state);
+                        *guard = TransactionInner::Fused(into.inner.clone());
+                        break;
+                    }
+                    TransactionInner::Fused(ref into) => into.clone(),
+                }
+            };
+            next = tmp;
+        }
+        // fuse our surfaces into our new transaction state
+        self.with_inner_state(|state| {
+            for (s, id) in my_state.surfaces {
+                if let Some(place) = state.surfaces.iter_mut().find(|place| place.0 == s) {
+                    // the surface is already in the list, update the serial
+                    if place.1 < id {
+                        place.1 = id;
+                    }
+                } else {
+                    // the surface is not in the list, insert it
+                    state.surfaces.push((s, id));
+                }
+            }
+            state.blockers.extend(my_state.blockers);
+        });
```

Comment:
These 10 lines or so look like a complete copy of `insert_state`.
I think it would be cleaner to just call that and then do state.blockers, no?

Or may this be such a hot path, that we want to avoid the frequent locking, or can this happen in a multi-threaded contexts and needs to happen atomically?

```suggestion
        for (s, id) in my_state.surfaces {
            self.insert_state(s, id);
        }
        self.with_inner_state(|state| {
            state.blockers.extend(my_state.blockers);
        });
```

Reply:
Mostly I found it wasteful to lock and unlock the state in a loop.

Reply:
Okay, I re-organized this with an `insert()` method on `TransactionState`. Is that better?

Reply:
Nice solution! :+1: